src/test/resources/optimized-query-for-filter.adoc

```diff
@@ -0,0 +1,897 @@
+      name: "Halle Berry"
+    }
+    reviews_some:{
+      from: {
+        name: "Jessica Thompson"
+      }
+      rating_gt: 2
+    }
+    released_gt: 1900
+  }) {
+    title
+  }
+}
+----
+
+.Cypher params
+[source,json]
+----
+{
+  "movie_actor_some_name": "Halle Berry",
+  "movie_reviews_some_rating": 2,
+  "movie_reviews_some_from_name": "Jessica Thompson",
+  "movie_released_gt": 1900
+}
+----
+
+.Cypher
+[source,cypher]
+----
+MATCH (movie:`Movie`)
+WHERE movie.released > $movie_released_gt
+WITH movie
+
+MATCH (movie)<-[:`ACTED_IN`]-(movie_actor_some:`Person`)
+WHERE movie_actor_some.name = $movie_actor_some_name
+WITH DISTINCT movie
+
+MATCH (movie)<-[review:`REVIEWED`]-(movie_reviews_some_from:`Person`)
+WHERE review.rating > $movie_reviews_some_rating
+  AND  movie_reviews_some_from.name = $movie_reviews_some_from_name
+WITH DISTINCT movie
+
+RETURN DISTINCT movie { .title } AS movie
+----
+
+.Movie Example result
+|===
+|title
+
+|Cloud Atlas
+|===
+
+=== Filter at same level (some + every)
+
+.Query configuration
+[source,json,config=true]
+----
+{  "optimizedQuery": true }
+----
+
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+  movie(filter: {
+    actor_some:{
+      name: "Halle Berry"
+    }
+    reviewedBy_every:{
+      name: "Jessica Thompson"
+    }
+  }) {
+    title
+  }
+}
+
+----
+
+.Cypher params
+[source,json]
+----
+{
+  "movie_actor_some_name": "Halle Berry",
+  "movie_reviewedBy_every_name": "Jessica Thompson"
+}
+----
+
+.Cypher
+[source,cypher]
+----
+MATCH (movie:`Movie`)
+WITH movie
+
+MATCH (movie)<-[:`ACTED_IN`]-(movie_actor_some:`Person`)
+WHERE
+    movie_actor_some.name = $movie_actor_some_name
+WITH DISTINCT movie
+
+MATCH (movie)<-[:`REVIEWED`]-(movie_reviewedBy_every:`Person`)
+WHERE
+    movie_reviewedBy_every.name = $movie_reviewedBy_every_name
+WITH movie,
+    size((movie)<-[:`REVIEWED`]-(:`Person`)) AS movie_reviewedBy_every_total,
+    count(DISTINCT movie_reviewedBy_every) AS movie_reviewedBy_every_count
+WHERE
+    movie_reviewedBy_every_total = movie_reviewedBy_every_count
+WITH DISTINCT movie
+
+RETURN DISTINCT movie { .title } AS movie
+----
+
+.Movie Example result
+|===
+|title
+
+|Cloud Atlas
+|===
+
+
+=== Filter at same level (every)
+
+.Query configuration
+[source,json,config=true]
+----
+{  "optimizedQuery": true }
+----
+
+.GraphQL-Query
+[source,graphql]
+----
+{
+  movie(filter: {
+    directedBy_every: {
+      name: "Clint Eastwood"
+    }
+    reviewedBy_every:{
+      name: "Jessica Thompson"
+    }
+  }) {
+    title
+  }
+}
+----
+
+.Cypher params
+[source,json]
+----
+{
+  "movie_directedBy_every_name":  "Clint Eastwood",
+  "movie_reviewedBy_every_name": "Jessica Thompson"
+}
+----
+
+.Cypher
+[source,cypher]
+----
+MATCH (movie:`Movie`)
+WITH movie
+
+MATCH (movie)<-[:`DIRECTED`]-(movie_directedBy_every:`Person`)
+WHERE movie_directedBy_every.name = $movie_directedBy_every_name
+WITH movie,
+    size((movie)<-[:`DIRECTED`]-(:`Person`)) AS movie_directedBy_every_total,
```

Comment:
unfortunately this is not as efficient but probably necessary. as it has to do an expand

Reply:
if we would know that there are only Person nodes at the end of DIRECTED we can leave off the label and it becomes a getDegree operation

Reply:
This is why the user should always be aware what the `_every` filter means. So we log warnings if he uses it by not defining it explicitly.